src/Compilers/CSharp/Test/Emit/CodeGen/SwitchTests.cs

```diff
@@ -9718,5 +9718,1207 @@ .locals init (int V_0)
+      IL_0107:  ldarg.0
+      IL_0108:  ldc.r8     9.1
+      IL_0111:  blt        IL_01dd
+      IL_0116:  ldarg.0
+      IL_0117:  ldc.r8     10.1
+      IL_0120:  bgt.un.s   IL_0142
+      IL_0122:  ldarg.0
+      IL_0123:  ldc.r8     9.1
+      IL_012c:  beq.s      IL_019d
+      IL_012e:  ldarg.0
+      IL_012f:  ldc.r8     10.1
+      IL_0138:  beq        IL_01bd
+      IL_013d:  br         IL_01ef
+      IL_0142:  ldarg.0
+      IL_0143:  ldc.r8     11.1
+      IL_014c:  beq.s      IL_01c6
+      IL_014e:  ldarg.0
+      IL_014f:  ldc.r8     12.1
+      IL_0158:  beq.s      IL_01a5
+      IL_015a:  br         IL_01ef
+      IL_015f:  ldarg.0
+      IL_0160:  ldc.r8     4.1
+      IL_0169:  bge.un.s   IL_0179
+      IL_016b:  ldarg.0
+      IL_016c:  ldc.r8     2.1
+      IL_0175:  bge.s      IL_01a1
+      IL_0177:  br.s       IL_01ef
+      IL_0179:  ldarg.0
+      IL_017a:  ldc.r8     5.1
+      IL_0183:  bge.s      IL_01eb
+      IL_0185:  ldarg.0
+      IL_0186:  ldc.r8     4.1
+      IL_018f:  beq.s      IL_01ca
+      IL_0191:  br.s       IL_01ef
+      IL_0193:  ldc.i4.s   19
+      IL_0195:  stloc.0
+      IL_0196:  br.s       IL_01f1
+      IL_0198:  ldc.i4.s   18
+      IL_019a:  stloc.0
+      IL_019b:  br.s       IL_01f1
+      IL_019d:  ldc.i4.5
+      IL_019e:  stloc.0
+      IL_019f:  br.s       IL_01f1
+      IL_01a1:  ldc.i4.1
+      IL_01a2:  stloc.0
+      IL_01a3:  br.s       IL_01f1
+      IL_01a5:  ldc.i4.8
+      IL_01a6:  stloc.0
+      IL_01a7:  br.s       IL_01f1
+      IL_01a9:  ldc.i4.s   15
+      IL_01ab:  stloc.0
+      IL_01ac:  br.s       IL_01f1
+      IL_01ae:  ldc.i4.s   13
+      IL_01b0:  stloc.0
+      IL_01b1:  br.s       IL_01f1
+      IL_01b3:  ldc.i4.s   20
+      IL_01b5:  stloc.0
+      IL_01b6:  br.s       IL_01f1
+      IL_01b8:  ldc.i4.s   9
+      IL_01ba:  stloc.0
+      IL_01bb:  br.s       IL_01f1
+      IL_01bd:  ldc.i4.6
+      IL_01be:  stloc.0
+      IL_01bf:  br.s       IL_01f1
+      IL_01c1:  ldc.i4.s   10
+      IL_01c3:  stloc.0
+      IL_01c4:  br.s       IL_01f1
+      IL_01c6:  ldc.i4.7
+      IL_01c7:  stloc.0
+      IL_01c8:  br.s       IL_01f1
+      IL_01ca:  ldc.i4.2
+      IL_01cb:  stloc.0
+      IL_01cc:  br.s       IL_01f1
+      IL_01ce:  ldc.i4.s   11
+      IL_01d0:  stloc.0
+      IL_01d1:  br.s       IL_01f1
+      IL_01d3:  ldc.i4.s   16
+      IL_01d5:  stloc.0
+      IL_01d6:  br.s       IL_01f1
+      IL_01d8:  ldc.i4.s   12
+      IL_01da:  stloc.0
+      IL_01db:  br.s       IL_01f1
+      IL_01dd:  ldc.i4.4
+      IL_01de:  stloc.0
+      IL_01df:  br.s       IL_01f1
+      IL_01e1:  ldc.i4.s   17
+      IL_01e3:  stloc.0
+      IL_01e4:  br.s       IL_01f1
+      IL_01e6:  ldc.i4.s   14
+      IL_01e8:  stloc.0
+      IL_01e9:  br.s       IL_01f1
+      IL_01eb:  ldc.i4.3
+      IL_01ec:  stloc.0
+      IL_01ed:  br.s       IL_01f1
+      IL_01ef:  ldc.i4.0
+      IL_01f0:  stloc.0
+      IL_01f1:  ldloc.0
+      IL_01f2:  ret
+    }
+"
+            );
```

Comment:
Verifying the IL is hard. Is there a way we could verify execution of different tests in some other way?
For example, could we define our own decimal type with custom operators that print out? Alternatively, we could include a Log operation in the `when` clauses?

Reply:
A `when` clause would prevent this machinery from kicking in, and would force things to be executed sequentially.
I think we have adequate tests for correctness, especially given the fuzz test `PatternMatchingTests3.RelationalFuzz_01`.
What is harder is verifying that the generated code uses *O(log N)* tests rather than *O(n)* tests.
However, I will add assertions to ValueDispatchNode that its height is less than *2 log N*, which is equivalent.

---
In reply to: [400461457](https://github.com/dotnet/roslyn/pull/42313#discussion_r400461457) [](ancestors = 400461457)